### ISSUE TITLE
Spec docs for opening up server side implementation for other languages

### ISSUE THREAD

**Issue by maxpert** [phretaddin/schemapack#16]
Is there a specs document that I am missing here? This will open up possibility for other developers to implement serializer/deserializer system in other languages. I for example want to do a Golang implementation.

**Comment by phretaddin** [comment]
No, you're not missing anything! Honestly, I never thought this would get as much interest as it did, being my first open source project and all, so I didn't think I needed one. However, not to worry: the specification is very simple. I'll explain it all right here!

First of all, the given schemas [are sorted recursively](https://github.com/phretaddin/schemapack/blob/master/schemapack.js#L289). The reason for this is because JavaScript objects have no iteration order. Without it, it is possible that schemas declared on the client and server COULD be iterated in a different order. Therefore, we sort them to give a deterministic traversal order. Strings are UTF8 by default, but have many different options available (mostly just because it was free because the underlying buffer module I was using already supported all of them).

Another concept to be familiar with is that of [variable length quantities](https://en.wikipedia.org/wiki/Variable-length_quantity). Basically, they're numbers where the last bit is used to tell if there are more bytes to come. Therefore, only 7 bits per byte are used to describe the value. These are used as length prefixes for arrays and strings. Arrays are prefixed with the item count as a `varuint` and strings are prefixed with the byte count as a `varuint`.

They're used because most of the time strings and arrays are short, so in the cases of their lengths being between 0 and 127, only one byte is needed to encode the length. When the lengths are between 128 to 16383, two bytes are needed. This lets us save a lot of space! There is also the datatype "varuint", which is just one of these variable length quantities, except as a value instead of a length prefix. Then we have "varint", which uses zig-zag encoding. If you're confused about any of this, just take a look at the source. All of these functions are pretty small and can transfer pretty easily between languages since it's just bit twiddling.

Other than those two concepts, the rest is pretty straightforward. You recursively iterate through the given object and schema, and write out the data types to the buffer. They're written in Big Endian, because the library is mostly used over networks and that is Network Byte Order.

There is no padding, special bytes, etc. It's similar to a C struct. No keys or delimiters need to be written to the buffer because you can just reference those in the schema while you're iterating through the encoding or decoding process.

Also, you can actually _read_ the buffer if you want, and that might help you understand the format as well. Try encoding a simple object and schema and going through and matching the bytes. It will quickly make complete sense. Here's an example straight from the README (but sorted already):

``` js
var playerSchema = sp.build({
  attributes: { agi: 'uint8', int: 'uint8', str: 'uint8' },
  health: 'varuint',
  jumping: 'boolean',
  position: [ 'int16' ],
});

var player = {
  attributes: { agi: 42, int: 22, str: 87 },
  health: 4000,
  jumping: false,
  position: [ -540, 343, 1201 ],
};

const buffer = playerSchema.encode(player);
console.log(buffer);
```

Now, let's take a look at that buffer:

`<Buffer 2a 16 57 a0 1f 00 03 fd e4 01 57 04 b1>`

Ok, it's in hex by default. Let's translate that to decimal to make it easier to read:

`<Buffer 42 22 87 160 31 00 03 253 228 01 87 04 177>`

Right off the bat we can see the `agi`, `int`, and `str`. Health is next. This is a varuint so it's difficult to translate by hand. We know it's two bytes. Let's try converting them both to binary:

`160 = 1010 0000`
`31   = 0001 1111`

The 1 in the first byte means to keep going. That is, include the next byte in the value.

The 0 in the second byte means that's it! We're done. Overlay the two 7-bit pairs next to each other in reverse order and you get:

`001 1111 010 0000`

Convert it to decimal and.. 4000!

The next byte is 0. That's simple. False is 0 and true is 1. The value is false.

The next byte is 3. That's the length prefix for the arrays. It means there's going to be 3 elements in the array.

The remaining items are all `int16` big endian signed integers. Two bytes each.

Hopefully that helped! Gotta run now but I'll be back later if you have any more questions. Let me know if you need a more detailed spec or have any questions. There's not much to it though! It pretty much just describes your data in as few bytes as possible.

**Comment by phretaddin** [comment]
Hey, did this help at all or do you need something more formal written up?

**Comment by NoNameProvided** [comment]
This should make it to its own file or in the wiki! Otherwise it will be hard to find.

**Comment by maxpert** [comment]
Well it helped a-lot. I am trying few things first before I make noise about it.

**Comment by nullpilot** [comment]
Hey. I am currently dabbling with a port of Schemapack to Elixir, and there's one thing about arrays that I would like to confirm before implementing it.

Assuming I have the following schema and data:  

``` javascript
let schema = sp.build({
      arr: ["uint8"],
    });
let data = {
      arr: [16, 17, 18, 19],
    };
```

The output would in bytes would be `00000100 00010000 00010001 00010010 00010011` - being the number of items as varuint (four), followed by the four values. All good.

If I add a second type to the schema array as follows:

``` javascript
let schema = sp.build({
      arr: ["uint8", "uint8"],
    });
```

The output changes to this: `00010000 00000100 00010001 00010010 00010011` - that is the first, static value, followed by the length and then the three values that belong to the repeatable last item. The order is entirely what I expected. The thing that I was surprised by is that the length is still four, so the length of the whole array, instead of changing to 3 for the length of the variable last part.

Can I rely on this always being the case, or is there something I have to keep in mind?

**Comment by crazyyi** [comment]
@phretaddin so a string and an array will both use one more byte for length prefix? 
My question below may not have much to do with this topic but I don't want to create another topic for it. For example, if I want to send a RGB color over the wire, if I send as an array `color: [rValue, gValue, bValue] `it will use 4 bytes in total in schemapack. However, if I send the color's RGB separately as the object's fields, such as `color: {r: rValue, g: gValue, b: bValue}` it will only use 3 bytes. Hopefully I am correct.

**Comment by nullpilot** [comment]
@crazyyi I'm not the maintainer, but since phretaddin wasn't active for a couple months now, I'll just answer this.

In short, yes. As an object you come out with a 3 byte binary, while with an array it will be 4 bytes long.

Example Object:
```javascript
let schema = sp.build({ r: "uint8", g: "uint8", b: "uint8" });
let data = {r: 255, g: 255, b: 255};
schema.encode(data);
// Output
<Buffer ff ff ff>
```

Example Array:
```javascript
let schema = sp.build(["uint8"]);
let data = [255, 255, 255];
schema.encode(data);
// Output
<Buffer 03 ff ff ff>
```

The reason for this is that there are no fixed length arrays, as the last given type in an array schema can always have any number of items of data later on.

**Comment by crazyyi** [comment]
Did you end up writing a golang implementation? If not I will have to roll out my own.

**Comment by maxpert** [comment]
@crazyyi go ahead would love to contribute.
